stdlib/public/core/Sort.swift

```diff
@@ -249,310 +248,456 @@ extension MutableCollection where Self: RandomAccessCollection {
+    bufferHigh = bufferLow + highCount
+    
+    var destHigh = high
+    var srcHigh = mid
+    destLow = mid
+
+    // Each iteration moves the element that compares higher into `destHigh`,
+    // preferring the buffer when equal to maintain stability. Elements are
+    // moved from either `bufferHigh - 1` or `srcHigh - 1`, with those
+    // pointers decrementing as elements are moved.
+    //
+    // Note: At the start of each iteration, each `...High` pointer points one
+    // past the element they're referring to.
+    while bufferHigh > bufferLow && srcHigh > low {
+      destHigh -= 1
+      if try areInIncreasingOrder(
+        (bufferHigh - 1).pointee, (srcHigh - 1).pointee
+      ) {
+        srcHigh -= 1
+        destHigh.moveInitialize(from: srcHigh, count: 1)
+        
+        // Moved an element from the lower initialized portion to the upper,
+        // sorted, initialized portion, so `destLow` moves down one.
+        destLow -= 1
+      } else {
+        bufferHigh -= 1
+        destHigh.moveInitialize(from: bufferHigh, count: 1)
       }
+    }
+  }
 
-    case (false, true):
-      // 1 swap: 132, 121 --- 2 swaps: 231, 221
-      // swap(b, c): 132->123, 121->112, 231->213, 221->212
-      swapAt(b, c)
+  return true
+}
 
-      if try areInIncreasingOrder(self[b], self[a]) {
-        // 213 (started as 231), 212 (started as 221)
-        // swap(a, b): 213->123, 212->122
-        swapAt(a, b)
-      }
+/// Calculates an optimal minimum run length for sorting a collection.
+///
+/// "... pick a minrun in range(32, 65) such that N/minrun is exactly a power
+/// of 2, or if that isn't possible, is close to, but strictly less than, a
+/// power of 2. This is easier to do than it may sound: take the first 6 bits
+/// of N, and add 1 if any of the remaining bits are set."
+/// - From the Timsort introduction, at
+///   https://svn.python.org/projects/python/trunk/Objects/listsort.txt
+///
+/// - Parameter c: The number of elements in a collection.
+/// - Returns: If `c <= 64`, returns `c`. Otherwise, returns a value in
+///   `32...64`.
+@inlinable
+internal func _minimumMergeRunLength(_ c: Int) -> Int {
+  // Max out at `2^6 == 64` elements
+  let bitsToUse = 6
+  
+  if c < 1 << bitsToUse {
+    return c
+  }
+  let offset = (MemoryLayout<Int>.size * 8 - bitsToUse) - c.leadingZeroBitCount
+  let mask = (1 << offset) - 1
+  return c >> offset + (c & mask == 0 ? 0 : 1)
+}
+
+/// Returns the end of the next in-order run along with a Boolean value
+/// indicating whether the elements in `start..<end` are in increasing order.
+@inlinable
+internal func _findNextRun<C: RandomAccessCollection>(
+  in elements: C,
+  from start: C.Index,
+  by areInIncreasingOrder: (C.Element, C.Element) throws -> Bool
+) rethrows -> (end: C.Index, ascending: Bool) {
+  var next = start
+  if next < elements.endIndex {
+    var current = next
+    elements.formIndex(after: &next)
+    if try next < elements.endIndex &&
```

Comment:
Can we extract this `next < elements.endIndex` as a guard, and then replace the `else if` few lines below with just an `else`?